Simplenote/src/main/java/com/automattic/simplenote/PreferencesFragment.java

```diff
@@ -158,6 +154,88 @@ public boolean onPreferenceChange(Preference preference, Object o) {
         });
     }
 
+    private static class SignOutAsyncTask extends AsyncTask<Void, Void, Boolean> {
+        private WeakReference<PreferencesFragment> fragmentWeakReference;
+
+        SignOutAsyncTask(PreferencesFragment fragment) {
+            fragmentWeakReference = new WeakReference<>(fragment);
+        }
+
+        @Override
+        protected Boolean doInBackground(Void... voids) {
+            PreferencesFragment fragment = fragmentWeakReference.get();
+            return fragment == null || fragment.hasUnsyncedNotes();
+        }
+
+        @Override
+        protected void onPostExecute(Boolean hasUnsyncedNotes) {
+            PreferencesFragment fragment = fragmentWeakReference.get();
+            if (fragment == null) {
+                return;
+            }
+
+            // Safety first! Check if any notes are unsynced and warn the user if so.
+            if (hasUnsyncedNotes) {
+                AlertDialog.Builder builder = new AlertDialog.Builder(fragment.getContext());
+                builder.setTitle(R.string.unsynced_notes);
+                builder.setMessage(R.string.unsynced_notes_message);
+                builder.setPositiveButton(R.string.delete_notes, fragment.signOutClickListener);
+                builder.setNeutralButton(R.string.visit_web_app, fragment.loadWebAppClickListener);
+                builder.setNegativeButton(R.string.cancel, null);
+                builder.show();
+            } else {
+                fragment.signOut();
+            }
+        }
+    }
+
+    private DialogInterface.OnClickListener signOutClickListener = new DialogInterface.OnClickListener() {
+        @Override
+        public void onClick(DialogInterface dialogInterface, int i) {
+            signOut();
+        }
+    };
+
+    private DialogInterface.OnClickListener loadWebAppClickListener = new DialogInterface.OnClickListener() {
+        @Override
+        public void onClick(DialogInterface dialogInterface, int i) {
+            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(WEB_APP_URL)));
+        }
+    };
+
+    public boolean hasUnsyncedNotes() {
```

Comment:
can be `private`. It will be passed to the asyncTask since it's a subclass of this class

Reply:
Also, move the subclass `SignOutAsyncTask` to be the last thing in the file (after `onUserCreated` method)